File: src/ipc.c
#include "ipc.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <sys/mman.h>
#include <sys/shm.h>
#include <assert.h>

struct shm_message {
    int pid;
    int kill;
};

static struct shm_message *message = NULL;
static char *shm_identifier = NULL;

static void char_replace(char *str, char oldChar, char newChar) {
    assert(str);
    char *strPtr = str;
    while ((strPtr = strchr(strPtr, oldChar)) != NULL)
        *strPtr++ = newChar;
}

void alloc_shared_memory(char *device_name, int button) {
    char *sanitized_device_name = strdup(device_name ? device_name : "");
    if (sanitized_device_name) {
        char_replace(sanitized_device_name, '/', '%');
    }

    int bytes = asprintf(&shm_identifier, "/mygestures_uid_%d_dev_%s_button_%d", getuid(),
                         sanitized_device_name ? sanitized_device_name : "", button);

    if (sanitized_device_name) {
        free(sanitized_device_name);
    }

    int shared_seg_size = sizeof(struct shm_message);
    int shmfd = shm_open(shm_identifier, O_CREAT | O_RDWR, 0600);

    if (shmfd < 0) {
        perror("In shm_open()");
        exit(shmfd);
    }
    int err = ftruncate(shmfd, shared_seg_size);

    message = (struct shm_message *)mmap(NULL, shared_seg_size,
                                         PROT_READ | PROT_WRITE, MAP_SHARED, shmfd, 0);

    if (message == MAP_FAILED) {
        perror("In mmap()");
        close(shmfd);
        exit(1);
    }
    close(shmfd);
}

void release_shared_memory() {
    if (shm_identifier) {
        if (shm_unlink(shm_identifier) != 0 && errno != ENOENT) {
            perror("In shm_unlink()");
        }
        free(shm_identifier);
        shm_identifier = NULL;
    }

    if (message && message != MAP_FAILED) {
        munmap(message, sizeof(struct shm_message));
        message = NULL;
    }
}

void send_kill_message(char *device_name) {
    assert(message);
    if (message->pid > 0) {
        printf("Asking mygestures running on pid %d to exit..\n", message->pid);
        int running = message->pid;
        message->pid = getpid();
        message->kill = 1;
        kill(running, SIGINT);
        usleep(100 * 1000); // 100ms
    }
    message->pid = getpid();
    message->kill = 0;
}

void on_interrupt(int a) {
    if (message && message != MAP_FAILED && message->kill) {
        printf("\nMygestures on PID %d asked me to exit.\n", message->pid);
    } else {
        printf("\nReceived the interrupt signal.\n");
        release_shared_memory();
    }
    exit(0);
}

void on_kill(int a) {
    release_shared_memory();
    exit(0);
}

File: src/ipc.rs
use std::ffi::CString;
use std::ptr;

#[repr(C)]
struct ShmMessage {
    pid: i32,
    kill: i32,
}

pub struct DaemonIpc {
    shm_name: String,
    ptr: *mut ShmMessage,
}

impl DaemonIpc {
    pub fn new(device_name: &str) -> Result<Self, std::io::Error> {
        let sanitized = device_name.replace('/', "%");
        let shm_name = format!(
            "/mygestures_uid_{}_dev_{}",
            unsafe { libc::getuid() },
            sanitized
        );
        let c_name = CString::new(shm_name.clone()).unwrap();

        let fd = unsafe { libc::shm_open(c_name.as_ptr(), libc::O_CREAT | libc::O_RDWR, 0o600) };
        if fd < 0 {
            return Err(std::io::Error::last_os_error());
        }

        let size = std::mem::size_of::<ShmMessage>();
        if unsafe { libc::ftruncate(fd, size as libc::off_t) } < 0 {
            let err = std::io::Error::last_os_error();
            unsafe {
                libc::close(fd);
            }
            return Err(err);
        }

        let ptr = unsafe {
            libc::mmap(
                ptr::null_mut(),
                size,
                libc::PROT_READ | libc::PROT_WRITE,
                libc::MAP_SHARED,
                fd,
                0,
            )
        };
        unsafe {
            libc::close(fd);
        }

        if ptr == libc::MAP_FAILED {
            return Err(std::io::Error::last_os_error());
        }

        Ok(DaemonIpc {
            shm_name,
            ptr: ptr as *mut ShmMessage,
        })
    }

    pub fn send_kill_message(&self) {
        unsafe {
            let msg = &mut *self.ptr;
            if msg.pid > 0 {
                println!("Asking mygestures running on pid {} to exit..", msg.pid);
                let running_pid = msg.pid;
                msg.pid = libc::getpid();
                msg.kill = 1;
                let _ = libc::kill(running_pid, libc::SIGINT);
                std::thread::sleep(std::time::Duration::from_millis(100));
            }
            msg.pid = libc::getpid();
            msg.kill = 0;
        }
    }

    pub fn is_kill_requested(&self) -> bool {
        unsafe {
            let msg = &*self.ptr;
            msg.kill != 0
        }
    }

    pub fn get_active_pid(&self) -> i32 {
        unsafe {
            let msg = &*self.ptr;
            msg.pid
        }
    }
}

impl Drop for DaemonIpc {
    fn drop(&mut self) {
        let size = std::mem::size_of::<ShmMessage>();
        unsafe {
            let _ = libc::munmap(self.ptr as *mut libc::c_void, size);
            let c_name = CString::new(self.shm_name.clone()).unwrap();
            let _ = libc::shm_unlink(c_name.as_ptr());
        }
    }
}
